crates/table/src/page_pool.rs

```diff
@@ -0,0 +1,294 @@
+use super::{indexes::Size, page::Page};
+use crate::indexes::max_rows_in_page;
+use crate::{page::PageHeader, MemoryUsage};
+use core::sync::atomic::{AtomicUsize, Ordering};
+use crossbeam_queue::ArrayQueue;
+use spacetimedb_sats::bsatn::{self, DecodeError};
+use spacetimedb_sats::de::{
+    DeserializeSeed, Deserializer, Error, NamedProductAccess, ProductVisitor, SeqProductAccess,
+};
+use std::sync::Arc;
+
+/// A page pool of currently unused pages available for use in [`Pages`](super::pages::Pages).
+#[derive(Clone)]
+pub struct PagePool {
+    inner: Arc<PagePoolInner>,
+}
+
+impl MemoryUsage for PagePool {
+    fn heap_usage(&self) -> usize {
+        let Self { inner } = self;
+        inner.heap_usage()
+    }
+}
+
+/// The default page pool has a size of 8 GiB.
+impl Default for PagePool {
+    fn default() -> Self {
+        Self::new(None)
+    }
+}
+
+impl PagePool {
+    /// Returns a new page pool with `max_size` bytes rounded down to the nearest multiple of 64 KiB.
+    ///
+    /// if no size is provided, a default of 8 GiB is used.
+    pub fn new(max_size: Option<usize>) -> Self {
+        const DEFAULT_MAX_SIZE: usize = 8 * (1 << 30); // 8 GiB
+        const PAGE_SIZE: usize = 64 * (1 << 10); // 64 KiB, `size_of::<Page>()`
+
+        let queue_size = max_size.unwrap_or(DEFAULT_MAX_SIZE) / PAGE_SIZE;
+        let pages = ArrayQueue::new(queue_size);
+        let inner = Arc::new(PagePoolInner {
+            pages,
+            unpooled_pages_count: <_>::default(),
+            dropped_pages_count: <_>::default(),
+        });
+        Self { inner }
+    }
+
+    /// Puts back a [`Page`] into the pool.
+    pub fn put(&self, page: Box<Page>) {
+        self.inner.put(page);
+    }
+
+    /// Puts back a [`Page`] into the pool.
+    pub fn put_many(&self, pages: impl Iterator<Item = Box<Page>>) {
+        for page in pages {
+            self.put(page);
+        }
+    }
+
+    /// Takes a [`Page`] from the pool or creates a new one.
+    ///
+    /// The returned page supports fixed rows of size `fixed_row_size`.
+    pub fn take_with_fixed_row_size(&self, fixed_row_size: Size) -> Box<Page> {
+        self.inner.take_with_fixed_row_size(fixed_row_size)
+    }
+
+    /// Takes a [`Page`] from the pool or creates a new one.
+    ///
+    /// The returned page supports a maximum of `max_rows_in_page` rows.
+    fn take_with_max_row_count(&self, max_rows_in_page: usize) -> Box<Page> {
+        self.inner.take_with_max_row_count(max_rows_in_page)
+    }
+
+    /// Deserialize a page from `buf` but reuse the allocations in the pool.
+    pub fn take_deserialize_from(&self, buf: &[u8]) -> Result<Box<Page>, DecodeError> {
+        self.deserialize(bsatn::Deserializer::new(&mut &*buf))
+    }
+
+    /// Returns the number of pages outside the pool.
+    ///
+    /// Note that if a page is dropped outside the pool,
+    /// it won't be aware and will count that as a leak.s
+    pub fn unpooled_pages_count(&self) -> usize {
```

Comment:
Isn't this going to happen any time we delete a database? This seems more likely to cause confusion than be helpful.

I think we track some metrics in the places where these pages are held, so we could probably get an accurate metric another way.